File: frontend/src/services/ingestionBatches.ts
/**
 * 统一 ingestion-batch API 封装
 */

import type { SseEvent } from '../lib/sse';
import { apiFetch } from './http';
import { openSseStream } from './sse';

export type ManifestSourceType = 'text' | 'url' | 'file';
export type BatchStatus =
  | 'queued'
  | 'running'
  | 'succeeded'
  | 'partial_failed'
  | 'failed'
  | 'canceled';
export type DocStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'canceled';

export interface ManifestTextEntry {
  source_type: 'text';
  entry_id?: string;
  title?: string;
  text: string;
}

export interface ManifestUrlEntry {
  source_type: 'url';
  entry_id?: string;
  title?: string;
  url: string;
}

export interface ManifestFileEntry {
  source_type: 'file';
  entry_id?: string;
  title?: string;
  material_id: string;
}

export type ManifestEntry = ManifestTextEntry | ManifestUrlEntry | ManifestFileEntry;

export interface IngestionBatchCreateRequest {
  kb_id: string;
  entries: ManifestEntry[];
}

export interface EntryError {
  entry_id: string;
  source_type: ManifestSourceType;
  code: string;
  message: string;
  retryable: boolean;
  details?: Record<string, unknown> | null;
}

export interface IngestionBatchSubmitResponse {
  batch_id: string;
  kb_id: string;
  status: BatchStatus;
  is_bootstrap: boolean;
  config_snapshot_id: string;
  config_version: number;
  total_docs: number;
  accepted_docs: number;
  failed_docs: number;
  entry_errors: EntryError[];
}

export interface IngestionBatchDoc {
  id: string;
  source_type: ManifestSourceType;
  source_ref: string | null;
  title: string | null;
  fingerprint: string;
  status: DocStatus;
  error_code: string | null;
  error_message: string | null;
  retry_count: number;
  retryable: boolean;
  chunk_count: number;
  config_version: number;
  created_at: string;
  updated_at: string;
}

export interface IngestionBatch {
  id: string;
  kb_id: string;
  config_snapshot_id: string;
  config_version: number;
  is_bootstrap: boolean;
  status: BatchStatus;
  total_docs: number;
  succeeded_docs: number;
  failed_docs: number;
  canceled_docs: number;
  succeeded_chunks: number;
  progress_percent: number;
  error_summary: Record<string, unknown> | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  docs: IngestionBatchDoc[];
}

export interface IngestionBatchRetryResponse {
  batch_id: string;
  status: BatchStatus;
  requeued_docs: number;
  ignored_docs: number;
}

export interface IngestionBatchCancelResponse {
  batch_id: string;
  status: BatchStatus;
  canceled_docs: number;
  finished_at: string | null;
}

export async function createIngestionBatch(
  data: IngestionBatchCreateRequest
): Promise<IngestionBatchSubmitResponse> {
  return apiFetch<IngestionBatchSubmitResponse>('/api/v1/ingestion-batches', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function getLatestIngestionBatch(kbId: string): Promise<IngestionBatch | null> {
  const query = new URLSearchParams({ kb_id: kbId, prefer_active: 'true' }).toString();
  const payload = await apiFetch<IngestionBatch | null>(`/api/v1/ingestion-batches/latest?${query}`);
  return payload ?? null;
}

export async function getIngestionBatch(batchId: string): Promise<IngestionBatch> {
  return apiFetch<IngestionBatch>('/api/v1/ingestion-batches/' + batchId);
}

export async function streamIngestionBatch(
  batchId: string,
  signal?: AbortSignal
): Promise<AsyncIterable<SseEvent>> {
  return openSseStream(`/api/v1/ingestion-batches/${batchId}/stream`, { method: 'GET' }, signal);
}

export async function retryIngestionBatch(batchId: string): Promise<IngestionBatchRetryResponse> {
  return apiFetch<IngestionBatchRetryResponse>('/api/v1/ingestion-batches/' + batchId + '/retry', {
    method: 'POST',
  });
}

export async function cancelIngestionBatch(batchId: string): Promise<IngestionBatchCancelResponse> {
  return apiFetch<IngestionBatchCancelResponse>('/api/v1/ingestion-batches/' + batchId + '/cancel', {
    method: 'POST',
  });
}

File: backend/src/app/schemas/ingestion_batches.py
"""统一 ingestion batch API 的 schema。"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class ManifestSourceType(str, Enum):
    TEXT = "text"
    URL = "url"
    FILE = "file"


class BatchStatus(str, Enum):
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELED = "canceled"


class DocStatus(str, Enum):
    QUEUED = "queued"
    PROCESSING = "processing"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELED = "canceled"


class ManifestTextEntry(BaseModel):
    source_type: Literal["text"] = ManifestSourceType.TEXT.value
    entry_id: str | None = Field(default=None, max_length=128)
    title: str | None = Field(default=None, max_length=500)
    text: str


class ManifestUrlEntry(BaseModel):
    source_type: Literal["url"] = ManifestSourceType.URL.value
    entry_id: str | None = Field(default=None, max_length=128)
    title: str | None = Field(default=None, max_length=500)
    url: str


class ManifestFileEntry(BaseModel):
    source_type: Literal["file"] = ManifestSourceType.FILE.value
    entry_id: str | None = Field(default=None, max_length=128)
    title: str | None = Field(default=None, max_length=500)
    material_id: uuid.UUID


ManifestEntry = Annotated[
    ManifestTextEntry | ManifestUrlEntry | ManifestFileEntry,
    Field(discriminator="source_type"),
]


class IngestionBatchCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kb_id: uuid.UUID
    entries: list[ManifestEntry] = Field(min_length=1)


class EntryErrorRead(BaseModel):
    entry_id: str
    source_type: ManifestSourceType
    code: str
    message: str
    retryable: bool
    details: dict | None = None


class IngestionBatchSubmitResponse(BaseModel):
    batch_id: uuid.UUID
    kb_id: uuid.UUID
    status: BatchStatus
    is_bootstrap: bool
    config_snapshot_id: uuid.UUID
    config_version: int
    total_docs: int
    accepted_docs: int
    failed_docs: int
    entry_errors: list[EntryErrorRead]


class IngestionBatchDocRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source_type: ManifestSourceType
    source_ref: str | None = None
    title: str | None = None
    fingerprint: str
    status: DocStatus
    error_code: str | None = None
    error_message: str | None = None
    retry_count: int
    retryable: bool
    chunk_count: int
    context_failed_chunks: list[dict] | None = None
    config_version: int
    created_at: datetime
    updated_at: datetime


class IngestionBatchRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kb_id: uuid.UUID
    config_snapshot_id: uuid.UUID
    config_version: int
    is_bootstrap: bool
    status: BatchStatus
    total_docs: int
    succeeded_docs: int
    failed_docs: int
    canceled_docs: int
    succeeded_chunks: int
    error_summary: dict | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    docs: list[IngestionBatchDocRead]


class KnowledgeBaseIngestionStateRead(BaseModel):
    kb_id: uuid.UUID
    has_active_batch: bool
    active_batch_id: uuid.UUID | None = None
    active_batch_status: BatchStatus | None = None
    updated_at: datetime


class IngestionBatchRetryResponse(BaseModel):
    batch_id: uuid.UUID
    status: BatchStatus
    requeued_docs: int
    ignored_docs: int


class IngestionBatchCancelResponse(BaseModel):
    batch_id: uuid.UUID
    status: BatchStatus
    canceled_docs: int
    finished_at: datetime | None = None
